### ISSUE TITLE
false-positive CS8631 and CS8634

### ISSUE THREAD

**Issue by KryuchkovAlexandr** [dotnet/roslyn#68773]
**Version Used**:  Visual Studio 2022 v17.6.4, .NET7

**Steps to Reproduce**:

```
public class Program
{
    interface IModel { }
    class ConcreteModel : IModel { }

    public void Method()
    {
        var sourceModel = new ConcreteModel();
        
        // CS8631 and CS8634 warnings
        Translate(sourceModel, out _);

        // no warnings
        Translate(sourceModel, out var unused);
    }

    private void Translate<T>(T source, out T translated) where T : class, IModel
    {
        translated = source;
    }
}
```

**Diagnostic Id**: CS8631, CS8634

**Expected Behavior**:
No warnings.

**Actual Behavior**:
Using discards triggers CS8631 and CS8634.

**Comment by PhilipCavanagh** [comment]
I just came here to report a very similar bug, but it seems to boil down to the same thing:

```
internal static class Program
{
    private static void Main()
    {        
        ThrowIfNull((string?)null, out _); // CS8634
    }
    
    private static void ThrowIfNull<T>(T? input, out T output) where T : class
    {
        output = input ?? throw new ArgumentNullException(nameof(input));
    }
}
```

- As above, this compiles if a variable is used instead of a discard.
- The example has the same error if `input` is changed to type `T`, and the argument changed to `""`.
- This compiles without issue on .NET SDKs 7.0.114 and 7.0.203 (latest patches I can find at time of writing).
- Fails to compile on 7.0.302 (first patch I can find) and later (and 8.0.100)
  - This corresponds to msbuild 17.6 and later
- I have a hunch this may be related to #66309 as that change is related to nullability, generics and discards, and was introduced as part of 17.6
